### DISCUSSION TITLE
cache-control

### DISCUSSION THREAD

**Discussion by Ivan-Chupin** [umputun/remark42#1549] [General]
Remark42 uses "cache-control: max-age=3600, no-cache" HTTP header for files:

1. embed.mjs
2. iframe.html
3. remark.mjs
4. remark.css 

Validators swear at this combination "max-age=3600, no-cache" and for good reason. It's kind of pointless. If we remove "max-age=3600" then absolutely nothing will change when browsers interact with the Remark42 application. Everything will work as before. If we remove "no-cache", then this will fundamentally change the caching mechanism. The browser will cache these files for one hour and will not even try to check them for changes on the server. This will increase performance and reduce traffic, but may cause problems when updating the above files.

**Comment by akellbl4** [comment]
> but may cause problems when updating the above files

This is the main problem here.
I'll try to think about our cache invalidation because there is other methods that could help us increase max age and remove `no-cache`.

**Reply by Ivan-Chupin** [reply]
The ideal caching solution would be to use unique URLs. For example. https​&#65279;://demo.remark42.com/web/**XXX**/remark.css or https​&#65279;://demo.remark42.com/web/remark.**XXX**.css where **XXX** is the version number. Then it would be possible to give browsers "**cache-control: max-age=31536000, immutable**". This would completely solve the problem with updates and make caching perfect.

**Reply by akellbl4** [reply]
using version doesn't make it perfect, since we can have files those weren't changed between releases. Name of files should include checksum so they will be reloaded only when they really changed without additional requests.

The problem of updating is still there. When you spin a new instance you're old one should stop at the same time if a user got html from old container they won't be able to get old static files from it, so that's the problem withc stops us from this upproatch.

**Reply by Ivan-Chupin** [reply]
In this case, before implementing proper caching, I suggest removing the "max-age=3600" directive from "cache-control: max-age=3600, no-cache" as inoperative and redundant.

**Reply by paskal** [reply]
I did that in https://github.com/umputun/remark42/pull/1592, please take a look.

**Reply by akellbl4** [reply]
>  The `no-cache` response directive indicates that the response can be stored in caches, but the response must be validated with the origin server before each reuse, even when the cache is disconnected from the origin server.

> Note that no-cache **does not mean "don't cache"**. no-cache allows caches to store a response but requires them to revalidate it before reuse. If the sense of "don't cache" that you want is actually "don't store", then no-store is the directive to use.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#no-cache

Maybe this description is misleading (or I understand it wrongly) which means we need to switch to `must-revalidate` instead of removing `max-age`

**Reply by paskal** [reply]
no-cache is more consistent with the behaviour of the current header and just clarifies then. Must-revalidate changes it to something which should be tested separately within Firefox, Chrome and Safari.
